src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -203,69 +251,241 @@ public object[] ArgumentList
+                        new ErrorRecord(
+                            new PSArgumentException(InternalCommandStrings.ParallelCannotUseTimeoutWithJob),
+                            "ParallelCannotUseTimeoutWithJob",
+                            ErrorCategory.InvalidOperation,
+                            this));
                 }
 
-                if (whatIf || confirm)
+                _taskJob = new PSTaskJob(
+                    Parallel.ToString(),
+                    ThrottleLimit);
+            }
+            else
+            {
+                _taskDataStreamWriter = new PSTaskDataStreamWriter(this);
+                _taskPool = new PSTaskPool(ThrottleLimit);
+                _taskPool.PoolComplete += (sender, args) =>
                 {
-                    string message = InternalCommandStrings.NoShouldProcessForScriptBlockSet;
-
-                    ErrorRecord errorRecord = new ErrorRecord(
-                        new InvalidOperationException(message),
-                        "NoShouldProcessForScriptBlockSet",
-                        ErrorCategory.InvalidOperation,
-                        null);
-                    ThrowTerminatingError(errorRecord);
+                    _taskDataStreamWriter.Close();
+                };
+                if (TimeoutSeconds != 0)
+                {
+                    _taskTimer = new System.Threading.Timer(
+                        (_) => _taskPool.StopAll(),
+                        null,
+                        TimeoutSeconds * 1000,
+                        System.Threading.Timeout.Infinite);
                 }
             }
+        }
 
-            // Calculate the start and end indexes for the processRecord script blocks
-            _end = _scripts.Count;
-            _start = _scripts.Count > 1 ? 1 : 0;
+        private void ProcessParallelParameterSet()
+        {
+            // Validate piped InputObject
+            if (_inputObject.BaseObject is ScriptBlock)
+            {
+                this.WriteError(
+                    new ErrorRecord(
+                            new PSArgumentException(InternalCommandStrings.ParallelPipedInputObjectCannotBeScriptBlock),
+                            "ParallelPipedInputObjectCannotBeScriptBlock",
+                            ErrorCategory.InvalidType,
+                            this));
 
-            // and set the end script if it wasn't explicitly set with a named parameter.
-            if (!_setEndScript)
+                return;
+            }
+
+            if (AsJob)
             {
-                if (_scripts.Count > 2)
-                {
-                    _end = _scripts.Count - 1;
-                    _endScript = _scripts[_end];
-                }
+                var taskChildJob = new PSTaskChildJob(
+                    Parallel,
+                    _usingValuesMap,
+                    InputObject);
+
+                _taskJob.AddJob(taskChildJob);
             }
+            else
```

Comment:
`else` -> `return`

Reply:
I don't know what these comments mean.

Reply:
Since there is no code following the `else` block, it may be simpler to just have a `return` statement after the `_taskJob.AddJob(taskChildJob);` line, and then the `else` can be removed, with the rest of the code remaining in the method body.

Reply:
I don't see any need for an early return change.  Seems like a matter of preference.

Reply:
Definitely preference / style. From what I've heard, the argument is that the unnecessary extra nesting is simply wasted characters and visually more complicated to parse.

I'd only really call it an issue if the `else` block is a good half-page or so long, though. 🙂